### PR TITLE
[exporter/clickhouse] Add component

### PR DESCRIPTION
**Pull request by hanjm**
**Description:** <Describe what has changed.>

Add a new exporter to export Opentelemetry logs to ClickHouse. It will also support spans and metrics in the future.
sponsor: @dmitryax 

**Link to tracking Issue:** <Issue number if applicable>
#2320

**Testing:** <Describe what testing was performed and which tests were added.>
This is the first PR to add a new component, it has only basic test code. according to the [guide](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#how-to-structure-prs-to-get-expedient-reviews).
**Documentation:** <Describe the documentation added.>
`exporter/clickhouseexporter/README.md`

This exporter supports sending OpenTelemetry logs to [ClickHouse](https://clickhouse.com/). It will also support spans and metrics in the future.
> ClickHouse is an open-source, high performance columnar OLAP database management system for real-time analytics using SQL.
> Throughput can be measured in rows per second or megabytes per second. 
> If the data is placed in the page cache, a query that is not too complex is processed on modern hardware at a speed of approximately 2-10 GB/s of uncompressed data on a single server.
> If 10 bytes of columns are extracted, the speed is expected to be around 100-200 million rows per second.

Note:
Always add [batch-processor](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/batchprocessor) to collector pipeline, as [ClickHouse document says:](https://clickhouse.com/docs/en/introduction/performance/#performance-when-inserting-data) 
> We recommend inserting data in packets of at least 1000 rows, or no more than a single request per second. When inserting to a MergeTree table from a tab-separated dump, the insertion speed can be from 50 to 200 MB/s.

#### User Cases

1. Use [Grafana Clickhouse datasource](https://grafana.com/grafana/plugins/grafana-clickhouse-datasource/) or
[vertamedia-clickhouse-datasource](https://grafana.com/grafana/plugins/vertamedia-clickhouse-datasource/) to make dashboard.
Support time-series graph, table and logs.

2. Analyze logs via powerful clickhouse SQL.

```clickhouse
/* get error count about my service last 1 hour.*/
SELECT count(*)
FROM logs
WHERE SeverityText="ERROR" AND timestamp >= NOW() - INTERVAL 1 HOUR;
/* find error log.*/
SELECT * 
FROM logs 
WHERE SeverityText="ERROR" AND timestamp >= NOW() - INTERVAL 1 HOUR;
/* find log with specific attribute .*/
SELECT Body
FROM logs 
WHERE Attributes.value[indexOf(Attributes.key, 'http.method')] = 'post' AND timestamp >= NOW() - INTERVAL 1 HOUR;
```

New component checklist: 

First PR:

- [x] Mention the sponsor in the PR description
- [x] A complete config.go, which will show how your component will look like to users
- [x] README.md with some basic information about your component. It doesn't have to be complete, but should contain at least:
- [x] Purpose and use-cases
- [x] Telemetry data types supported
- [x] versions.yaml
- [x] CODEOWNERS with at least the sponsor and the component's author
- [x] internal/components tests
- [x] go.mod at the root of the repository and cmd/configschema
- [x] dependabot
- [x] The skeleton of the component, potentially using the helpers
- [x] Makefile

### CODE REVIEW

**Comment by hanjm** [comment]
Hi @codeboten @morigs, Could you help me review? Thank you. 😄

File: exporter/clickhouseexporter/exporter_test.go
Comment (MovieStoreGuy) [inline exporter/clickhouseexporter/exporter_test.go:90]: This potentially could be problematic if you want to test that the exporter shut downs with/out error. I would suggest leaving it to the caller to ensure shutdown is called.

File: exporter/clickhouseexporter/exporter_test.go
Comment (MovieStoreGuy) [inline exporter/clickhouseexporter/exporter_test.go:86]: I would also advocate for using `testing.TB` instead of `*testing.T` so in future, you can use this method to create an export for either a test method or a benchmark method.

File: exporter/clickhouseexporter/factory.go
Comment (MovieStoreGuy) [inline exporter/clickhouseexporter/factory.go:21]: I am not sure I am okay with blank imports. Is there an expected side affect that is required here?

File: exporter/clickhouseexporter/factory.go
Comment (MovieStoreGuy) [inline exporter/clickhouseexporter/factory.go:66]: There is additional helpers that you can adopt here as well, I would suggest looking at the following ones:

- `exporterhelper.WithTimeout`
- `exporterhelper.WithRetry`

This would help make the exporter more durable / reliable for those whom adopt it.

File: go.mod
Comment (MovieStoreGuy) [inline go.mod]: This feels like an error, would you mind just double checking this?

File: exporter/clickhouseexporter/README.md
Comment (MovieStoreGuy) [reply exporter/clickhouseexporter/README.md:1]: I would also add that this is limited to logs only and should be considered experimental.

**Review by MovieStoreGuy — CHANGES_REQUESTED** [review]
There is a few things that I should point out and ensure that is followed here, it is strongly suggested that when adding a new component to the collector that you follow [adding new components guideline](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components)

Typically, it is expected that the first commit is the skeleton / outline of your component, then working example with details of configuration / readme, then the final PR is adding the component into the collector. 

I would consider this to be step one and two of what I said above done together, which isn't bad just makes the feedback process longer. 

From I see, I am mostly happy with structure of it, however some of the logic and error handling should be addressed to ensure those using it can provide useful feedback with as much information as possible.

I suggest while you wait for more feedback from the community that you branch of this and finish the steps listed in the guide to parallelise the feedback process.

**Comment by hanjm** [comment]
Thanks @MovieStoreGuy review, I will reduce the size of this PR, follow the guide.

File: exporter/clickhouseexporter/README.md
Comment (morigs) [inline exporter/clickhouseexporter/README.md:9]: Just something to think about, not a recommendation: it's possible to use [Buffer engine](https://clickhouse.com/docs/en/engines/table-engines/special/buffer/) for this purpose. But it has its own drawbacks.

File: exporter/clickhouseexporter/README.md
Comment (morigs) [inline exporter/clickhouseexporter/README.md:15]: HTTP interface is very common and a lot of people will assume that `address` means HTTP base URL. I suggest adding a note here about address format (and/or probably give a link to the clickhouse-go repo).

File: exporter/clickhouseexporter/README.md
Comment (morigs) [inline exporter/clickhouseexporter/README.md:89]: This looks like schema for traces, not logs. Correct me if I'm wrong

**Review by morigs — COMMENTED** [review]
It would be awesome to support all types of data (spans, metrics and logs). I believe Clickhouse is flexible and fast enough for this.
However, perhaps, this is for the future 😃

**Comment by hanjm** [comment]
> It would be awesome to support all types of data (spans, metrics and logs). I believe Clickhouse is flexible and fast enough for this.
> However, perhaps, this is for the future 😃

@morigs Strongly agree. It will be implement in the future, fisrt step i am implementing logs, it is the most sample across three types of data (spans, metrics, logs).

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [reply exporter/clickhouseexporter/README.md:89]: It is logs model, the `traceId` field is for correlate with trace spans. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition

Comment (hanjm) [reply exporter/clickhouseexporter/README.md:89]: I am also refence the elasticsearchexporter. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/elasticsearchexporter/model.go#L42

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [reply exporter/clickhouseexporter/README.md:15]: Thanks. You are right. let me add more comment for it.

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [reply exporter/clickhouseexporter/README.md:9]: Buffer is awsome feature, but it can not replace the batch-processor. 
> Note that it does not make sense to insert data one row at a time, even for Buffer tables. This will only produce a speed of a few thousand rows per second, while inserting larger blocks of data can produce over a million rows per second (see the section “Performance”).

the `Performance` section clickhouse document says that 
> We recommend inserting data in packets of at least 1000 rows, or no more than a single request per second. 

https://clickhouse.com/docs/en/introduction/performance/#performance-when-inserting-data.

Comment (hanjm) [reply exporter/clickhouseexporter/README.md]: I have update README.md just now.

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [reply exporter/clickhouseexporter/README.md]: Updated. I use raw dsn for more flexibility.

**Comment by hanjm** [comment]
Hi @codeboten @morigs @MovieStoreGuy, All the checks are passed 🚀 . Please review again 😃. Thank you.

File: exporter/clickhouseexporter/README.md
Comment (morigs) [reply exporter/clickhouseexporter/README.md]: Oh, ok. I was looking in the `pdata.Logs` but didn't manage to find this)

File: exporter/clickhouseexporter/README.md
Comment (morigs) [inline exporter/clickhouseexporter/README.md:78]: At least some of these fields should be nullable. There is an [example](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#example-log-records) of log containing timestamp and attributes only.
Because `pushLogsData` is currently empty, I'm not sure what the plan is.
Because nullable causes performance degradation, I assume your plan is inserting "default" values for empty fields. Am I right?

File: exporter/clickhouseexporter/config.go
Comment (morigs) [inline exporter/clickhouseexporter/config.go:33]: Do you have plans on supporting both?

Comment (hanjm) [reply exporter/clickhouseexporter/config.go:33]: yes, it all compatibility with database/sql, I think it it easy to support both. if there is difficult. i prefer use tcp.

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [reply exporter/clickhouseexporter/README.md:78]: yes. diffrent logs has various attributes, next plan we could make the table name configurable, so we can use [routing processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/routingprocessor) to write dirrent log type to dirrent table for better performance.

**Comment by jpkrohling** [comment]
@hanjm, could you please go over #6926 and see if this PR follows that? I would really insist in breaking this into multiple PRs to make it easier to review, but I believe this also needs a sponsor.

**Comment by hanjm** [comment]
@jpkrohling  Thank you . I will refactor my PR follow you detailed checklist. 👍 

First PR:

[] Mention the sponsor in the PR description
[x] A complete config.go, which will show how your component will look like to users
README.md with some basic information about your component. It doesn't have to be complete, but should contain at least:
[x] Purpose and use-cases
[x] Telemetry data types supported
[x] versions.yaml
[x] CODEOWNERS with at least the sponsor and the component's author
[] internal/components tests
[x] go.mod at the root of the repository and cmd/configschema
[ ] dependabot
[x] The skeleton of the component, potentially using the helpers
[x] Makefile

Could you help me find a sponsor?

**Comment by jpkrohling** [comment]
> Could you help me find a sponsor?

I pinged the approvers in the tracking issue. Let's see if one of them will volunteer, but please be patient, as most approvers are out due to the holiday season.

**Comment by jpkrohling** [comment]
@dmitryax has volunteered to sponsor this one, he'll be reviewing this one shortly.

File: exporter/clickhouseexporter/config.go
Comment (dmitryax) [inline exporter/clickhouseexporter/config.go:29]: `exporterhelper.RetrySettings` usually goes with `exporterhelper.QueueSettings` to control the queue size. Is there a reason to omit it?

**Review by dmitryax — COMMENTED** [review]
Just one Q. Otherwise LGTM. Looking forward to the next PRs

File: exporter/clickhouseexporter/config.go
Comment (hanjm) [reply exporter/clickhouseexporter/config.go:29]: As ClickHouse recommend inserting data in packets of at least 1000 rows, or no more than a single request per second. 
The `exporterhelper.QueueSettings` has a field `NumConsumers`, So I omit the `exporterhelper.QueueSettings` to keep it only one exporter goroutine, it consume the data from batch-processor, which batch by size 1000 or time 1s.  

I rethink it again, may be use a subset field `QueueSettings` will be better, only keep `queue_size`,  like awslogexporter does. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/awscloudwatchlogsexporter/config.go#L80

Comment (hanjm) [reply exporter/clickhouseexporter/config.go:29]: I have add `QueueSettings` field in latest commit (:

Comment (dmitryax) [reply exporter/clickhouseexporter/config.go:29]: Sounds good. thanks

File: exporter/clickhouseexporter/config.go
Comment (dmitryax) [inline exporter/clickhouseexporter/config.go:66]: This may be the reason for failing linter

Comment (hanjm) [reply exporter/clickhouseexporter/config.go]: Thank you, Fixed.

**Review by dmitryax — APPROVED** [review]

File: internal/components/components.go
Comment (jpkrohling) [inline internal/components/components.go:233]: Don't forget to open a PR with the releases repo to add this component there: https://github.com/open-telemetry/opentelemetry-collector-releases

**Review by jpkrohling — DISMISSED** [review]

File: exporter/clickhouseexporter/exporter.go
Comment (jpkrohling) [inline exporter/clickhouseexporter/exporter.go:58]: @dmitryax, shouldn't the first PR already include the implementation?
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components

Comment (jpkrohling) [reply exporter/clickhouseexporter/exporter.go:58]: True. I see the bullet points on that section as things that should go within the first PR, but I guess that part got left off. I'm fine with smaller PRs if that's your preference.

File: internal/components/components.go
Comment (hanjm) [reply internal/components/components.go:233]: OK

**Comment by hanjm** [comment]
Thank you for reviewing 🚀

**Comment by K-Prabha** [comment]
Do we have a timeline on when this capability will be available?

**Comment by hanjm** [comment]
@K-Prabha it will be available when https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8028 be merged.

**Pull request merged**
